src/phpDocumentor/ConfigurationFactory.php

```diff
@@ -0,0 +1,138 @@
+<?php
+
+namespace phpDocumentor;
+
+final class ConfigurationFactory
+{
+    /**
+     * @var Uri
+     */
+    private $uri;
+
+    /**
+     * @var \SimpleXMLElement
+     */
+    private $xml;
+
+    /**
+     * @var string
+     */
+    private $schemaPath;
+
+    /**
+     * @var bool
+     */
+    private $validateUri;
+
+    /**
+     * @param Uri    $uri
+     * @param string $schemaPath
+     */
+    public function __construct(Uri $uri, $schemaPath = '')
+    {
+        if ($schemaPath === '') {
+            $schemaPath = __DIR__ . '/../../data/xsd/phpdoc.xsd';
+        }
+
+        $this->replaceLocation($uri);
+        $this->schemaPath = $schemaPath;
+    }
+
+    /**
+     * Replaces the location of the configuration file if it is different.
+     *
+     * @param Uri $uri
+     */
+    public function replaceLocation(Uri $uri)
+    {
+        if ($this->uri !== $uri) {
+            $this->validateUri = true;
+        }
+
+        $this->uri = $uri;
+    }
+
+    /**
+     * Converts the phpDocumentor configuration xml to an array.
+     *
+     * @return array
+     */
+    public function get()
+    {
+        $this->validate($this->uri);
+
+        $version = $this->checkIfVersionAttributeIsPresent($this->xml);
+        if ($version) {
+            $this->validateXmlStructure($this->xml);
+            $xml = new Phpdoc3XmlToArrayConverter($this->xml);
```

Comment:
You have hardcoded the strategies but it is more convenient to inject which strategies there are in the constructor and in this location loop over every registered strategy and use a `match` method in a strategy class to check if the XML can be handled by that strategy; if it can be handled you can pass the xml to the `handle` method which does the actual conversion.

There are several advantages to injecting classes:

1. You can test them individually; now you need to test each of these classes using a unit test for the factory and hence you do not test a single unit
2. You can pass dependencies specific to a strategy; because you instantiate the strategy object outside of this class you can use varying constructor arguments and thus inject separate dependencies per strategy